src/middlewares/createMetadataMiddleware.ts

```diff
@@ -0,0 +1,104 @@
+import { InstantSearch, Middleware, Widget } from '../types';
+import { Index } from '../widgets/index/index';
+
+type WidgetMetaData = {
+  type?: string;
+  params: string[];
+};
+
+type Payload = {
+  widgets: WidgetMetaData[];
+};
+
+function extractPayload(
+  widgets: Array<Widget<{ renderState: any }>>,
+  instantSearchInstance: InstantSearch,
+  payload: Payload
+) {
+  const parent = instantSearchInstance.mainIndex;
+
+  const initOptions = {
+    instantSearchInstance,
+    parent,
+    scopedResults: parent.getScopedResults(),
+    state: parent.getHelper()!.state,
+    helper: parent.getHelper()!,
+    createURL: parent.createURL,
+    uiState: instantSearchInstance._initialUiState,
+    renderState: instantSearchInstance.renderState,
+    templatesConfig: instantSearchInstance.templatesConfig,
+    searchMetadata: {
+      isSearchStalled: instantSearchInstance._isSearchStalled,
+    },
+  };
+
+  widgets.forEach(widget => {
+    let widgetParams = {};
+
+    if (widget.getWidgetRenderState) {
+      const renderState = widget.getWidgetRenderState(initOptions);
+
+      if (renderState && renderState.widgetParams) {
+        widgetParams = renderState.widgetParams;
+      }
+    }
+
+    // since we destructure in all widgets, the parameters with defaults are set to "undefined"
+    const params = Object.keys(widgetParams).filter(
+      key => widgetParams[key] !== undefined
+    );
+
+    payload.widgets.push({
+      type: widget.$$type,
+      params,
+    });
+
+    if (widget.$$type === 'ais.index') {
+      extractPayload(
+        (widget as Index).getWidgets(),
+        instantSearchInstance,
+        payload
+      );
+    }
+  });
+}
+
+export function isMetadataEnabled() {
+  return (
+    typeof window !== 'undefined' &&
+    window.navigator.userAgent.indexOf('Algolia Crawler') > -1
+  );
+}
+
+export function createMetadataMiddleware(): Middleware {
```

Comment:
"metadata" is not very descriptive. How about `createMetadataDomMiddleware`?

Reply:
I can't find a better description, dom seems redundant to me. It's creating a `meta` tag (short for metadata) and filling it. Since there's no other metadata in InstantSearch, this seems correct to me